pytential/symbolic/execution.py

```diff
@@ -932,6 +934,51 @@ def __repr__(self):
     def __str__(self):
         return f"{type(self).__name__}({repr(self.places)})"
 
+
+_KNOWN_GEOMETRY_COLLECTION_CACHE_KEYS = (
+        EvaluationMapperCSECacheKey,
+        _GeometryCollectionConnectionCacheKey,
+        _GeometryCollectionDiscretizationCacheKey,
+        )
+
+
+def add_geometry_to_collection(
+        places: GeometryCollection,
+        geometries: Dict[Hashable, Any]) -> GeometryCollection:
+    """Adds a :class:`dict` of geometries to an existing collection.
+
+    This function is similar to :meth:`GeometryCollection.merge`, but it makes
+    an attempt to maintain the caches in *places*. In particular, a shallow
+    copy of the following are pased to the new collection
+
+    * Any cached discretizations from
+      :func:`~pytential.qbx.refinement.refine_geometry_collection`.
+    * Any cached expressions marked with `cse_scope.DISCRETIZATION` from the
+      evaluation mapper.
+
+    This allows adding new targets to the collection without recomputing the
+    source geometry data.
+    """
+    from pytential.source import PointPotentialSource
+    from pytential.target import PointsTarget
+    for key, geometry in geometries.items():
+        if key in places.places:
+            raise ValueError(f"geometry '{key}' already in the collection")
+
+        if not isinstance(geometry, (PointsTarget, PointPotentialSource)):
+            raise TypeError(
+                    f"Cannot add a geometry of type '{type(geometry).__name__}' "
+                    "to the existing collection. Construct a new collection "
+                    "instead.")
+
+    # copy over all the caches
+    new_places = places.merge(geometries)
+    for key in _KNOWN_GEOMETRY_COLLECTION_CACHE_KEYS:
```

Comment:
Why not iterate over all existing cache names and at least warn if one is encountered that's not understood?